addons/mail/controllers/thread.py

```diff
@@ -144,15 +144,15 @@ def _prepare_post_data(self, post_data, thread, partner_emails=None, **kwargs):
                 partner_emails,
                 no_create=not request.env.user.has_group("base.group_partner_manager"),
             )
-        if not request.env.user._is_internal():
-            partners = partners & self._filter_message_post_partners(thread, partners)
-        post_data["partner_ids"] = partners.ids
         if role_ids := post_data.pop("role_ids", []):
-            user_ids = request.env["res.users"].search([("role_ids", "in", role_ids)])
-            post_data["partner_ids"] += user_ids.partner_id.ids
+            # sudo - res.users: getting partners linked to the role is allowed.
+            partners |= request.env["res.users"].sudo().search([("role_ids", "in", role_ids)]).partner_id
+        post_data["partner_ids"] = self._filter_message_post_partners(thread, partners).ids
         return post_data
 
     def _filter_message_post_partners(self, thread, partners):
+        if request.env.user._is_internal():
+            return partners
```

Comment:
Using internal user is not correct, you might access an otherwise private document with a portal token even as an internal user.

Condition for follower based check should not be changed in this PR.

Reply:
As discussed it's unchanged actually I miss-read diff. Fix for taking this use case into account can be done in a separate PR.

Reply:
Or it can be done here directly, checking `sudo(False).has_access("write")` should be sufficient I believe?